File: 4.8-bc/py/stage-3-hexagonal/app/auth/app/service.py
"""Auth use cases — Register, Login.

Depends on the domain ``UserRepository`` Protocol and ``shared.events.EventBus``.
Knows nothing about HTTP or asyncpg.
"""
from __future__ import annotations

from datetime import UTC, datetime
from uuid import uuid4

from passlib.hash import bcrypt

from app.auth.domain.errors import (
    EmailAlreadyExistsError,
    InvalidCredentialsError,
    UserNotFoundError,
)
from app.auth.domain.repository import UserRepository
from app.auth.domain.user import User, UserRegistered
from app.shared.events import EventBus


class Service:
    def __init__(self, repo: UserRepository, bus: EventBus) -> None:
        self._repo = repo
        self._bus = bus

    async def register(self, email: str, password: str) -> User:
        password_hash = bcrypt.hash(password)
        u = User(
            id=uuid4(),
            email=email,
            password_hash=password_hash,
            created_at=datetime.now(UTC),
        )
        try:
            await self._repo.create(u)
        except EmailAlreadyExistsError:
            raise
        await self._bus.publish(
            UserRegistered(user_id=u.id, email=u.email, at=u.created_at)
        )
        return u

    async def login(self, email: str, password: str) -> User:
        try:
            u = await self._repo.find_by_email(email)
        except UserNotFoundError as exc:
            raise InvalidCredentialsError() from exc
        if u is None:
            raise InvalidCredentialsError()
        if not bcrypt.verify(password, u.password_hash):
            raise InvalidCredentialsError()
        return u

File: 4.8-bc/go/stage-3-hexagonal/auth/app/service.go
// Package app — use cases для Auth BC. Залежить тільки від domain interfaces.
package app

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgconn"
	"golang.org/x/crypto/bcrypt"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/auth/domain"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/events"
)

type Service struct {
	repo domain.UserRepository
	bus  *events.Bus
}

func New(repo domain.UserRepository, bus *events.Bus) *Service {
	return &Service{repo: repo, bus: bus}
}

func (s *Service) Register(ctx context.Context, email, password string) (*domain.User, error) {
	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	u := domain.User{
		ID:           uuid.New(),
		Email:        email,
		PasswordHash: string(hash),
		CreatedAt:    time.Now().UTC(),
	}

	if err := s.repo.Create(ctx, u); err != nil {
		var pgErr *pgconn.PgError
		if errors.As(err, &pgErr) && pgErr.Code == "23505" {
			return nil, domain.ErrEmailAlreadyExists
		}
		return nil, err
	}

	s.bus.Publish(ctx, domain.UserRegistered{UserID: u.ID, Email: u.Email, At: u.CreatedAt})
	return &u, nil
}

func (s *Service) Login(ctx context.Context, email, password string) (*domain.User, error) {
	u, err := s.repo.FindByEmail(ctx, email)
	if err != nil {
		return nil, domain.ErrInvalidCredentials
	}
	if err := bcrypt.CompareHashAndPassword([]byte(u.PasswordHash), []byte(password)); err != nil {
		return nil, domain.ErrInvalidCredentials
	}
	return u, nil
}
